### ISSUE TITLE
[Question] Do use_liger_gkd_loss breaks DDP gradient sync by unwrapping student model？

### ISSUE THREAD

**Issue by JeffLee1874** [huggingface/trl#5282]
In the GKD and GOLD trainers, when `use_liger_gkd_loss` is enabled, the code explicitly unwraps the student model before the forward pass:

```python
unwrapped_student = self.accelerator.unwrap_model(model)
if hasattr(unwrapped_student, "get_decoder") and unwrapped_student.get_decoder() is not None:
    base_student = unwrapped_student.get_decoder()
else:
    base_student = getattr(
        unwrapped_student, getattr(unwrapped_student, "base_model_prefix", "model"), unwrapped_student
    )

student_outputs = base_student(
    input_ids=inputs["input_ids"],
    attention_mask=inputs["attention_mask"],
    use_cache=False,
)
```
Isn't this problematic for distributed training? By bypassing the DistributedDataParallel wrapper for the student's forward pass, the gradients won't be synchronized across GPUs via AllReduce, causing the model weights on different ranks to diverge. Shouldn't the student model be called directly (e.g., model(...)) to preserve the DDP graph?

**Comment by Saad-Mallebhari** [comment]
@JeffLee1874 
Your concern is worth raising. In DDP, gradient synchronization relies on autograd hooks that are registered on parameters during the wrapped model's forward pass. If `base_student(...)` is called on the unwrapped model directly, those hooks may not be registered for that forward pass , which means AllReduce may not trigger correctly during backward, potentially causing weights to diverge across ranks.

That said, whether this actually breaks sync depends on a few things:
- If the DDP-wrapped `model(...)` is still called elsewhere in the same backward graph, the hooks may already be registered and sync could still occur
- If `base_student` is only used to extract hidden states for the Liger kernel and gradients don't flow back through it, the DDP concern may not apply in practice

A few things worth checking to confirm:
- Log parameter norm across ranks after a few steps with `use_liger_gkd_loss=True` , diverging norms would confirm sync is broken
- Check whether `model(...)` (the wrapped version) is called anywhere else in the same training step

The unwrap looks intentional for accessing internal model components needed by the Liger loss kernel, but if gradients do flow through `base_student`, the lack of DDP wrapping may silently affect training correctness in a multi-GPU setup.

**Comment by albertvillanova** [comment]
Hey @JeffLee1874, sorry for the slow reply — catching up on the backlog. 

Your instinct is right and the concern is real. DDP relies on its wrapper's `forward()` calling `prepare_for_backward()` to reset the reducer's per-iteration state; here we call the unwrapped decoder directly (https://github.com/huggingface/trl/blob/c3f21ede9afc0e2cadba183b3be376c5405cf65b/trl/experimental/gkd/gkd_trainer.py#L298-L310), so that bookkeeping is skipped. The AllReduce hooks attached at DDP init still fire on backward (they're on the param accumulators, not on `forward`), so in simple single-iteration setups gradients may still sync — but across iterations the reducer state drifts, which surfaces either as a hard `Expected to have finished reduction in the prior iteration...` error or, worse, silent misaccounting when `find_unused_parameters=True`. The right fix is to call the wrapped student with `output_hidden_states=True` and use `hidden_states[-1]` as the liger input, instead of bypassing DDP. We'll need the same fix in the GOLD/distillation/SDPO/SDFT experimental trainers, which all repeat this pattern. 

Thanks for the careful read.

**Comment by albertvillanova** [comment]
I proposed a fixing PR:
- https://github.com/huggingface/trl/pull/5934

**Comment by JeffLee1874** [comment]
> I proposed a fixing PR:
> 
> * [Fix: Route liger student forward through DDP wrapper in GKD, GOLD, and Distillation trainers #5934](https://github.com/huggingface/trl/pull/5934)

Thanks for your great work !
